common/djangoapps/student/models.py

```diff
@@ -1552,11 +1553,155 @@ def send_signal_full(cls, event, user=user, mode=mode, course_id=None, cost=None
                                   mode=mode, course_id=course_id,
                                   cost=cost, currency=currency)
 
-    def emit_event(self, event_name, enterprise_uuid=None):
+    @staticmethod
+    def _prepare_date_block(course_date_list, block, block_date, timezone_offset):
+        """
+        Prepare date block which include assignment related data fot this date
+        """
+        block = {
+            'title': block['title'] or '',
+            'assignment_type': block['assignment_type'] or '',
+            'link': block['link'] or '',
+            'assignment_count': (course_date_list['assignment_count'] + 1 if block['assignment_type']
+                                 else course_date_list['assignment_count']),
+            'due_date': block_date.strftime("%a, %b %d, %Y"),
+            'due_time': (f'{block_date.strftime("%H:%M %p")} GMT{timezone_offset}'
+                         if block['assignment_type'] else '')
+        }
+        return block
+
+    def get_course_dates_for_email(self, user, course_id, request):
+        from lms.djangoapps.courseware.courses import get_course_date_blocks, get_course_with_access
+        from lms.djangoapps.courseware.date_summary import TodaysDate
+        from lms.djangoapps.courseware.context_processor import user_timezone_locale_prefs
+        from lms.djangoapps.course_home_api.dates.serializers import DateSummarySerializer
+        """
+        Getting nearest dates from today one would be before today and one
+        would be after today.
+        """
+        user_timezone_locale = user_timezone_locale_prefs(request)
+        user_timezone = timezone(user_timezone_locale['user_timezone'] or str(UTC))
+
+        course = get_course_with_access(user, 'load', course_id, check_if_enrolled=False)
+        date_blocks = get_course_date_blocks(course, user, request, include_access=True, include_past_dates=True)
+        date_blocks = [block for block in date_blocks if not isinstance(block, TodaysDate)]
+        blocks = DateSummarySerializer(date_blocks, many=True).data
+
+        course_date = {
+            'date': '',
+            'title': '',
+            'assignment_type': '',
+            'link': '',
+            'assignment_count': -1,
+            'due_date': '',
+            'due_time': ''
+        }
+        course_date_list = [
+            {
+                'id': 1,
+                **course_date,
+            },
+            {
+                'id': 2,
+                **course_date,
+            },
+            {
+                'id': 3,
+                **course_date,
+            }
+        ]
+        today = datetime.now(UTC).date()
+
+        for block in blocks:
+            block_date = datetime.strptime(block['date'][:19], '%Y-%m-%dT%H:%M:%S')
+            block_date = block_date.replace(tzinfo=UTC).date()
+
+            if block_date < today:
+                course_date_list[0]['date'] = block_date
```

Comment:
Are we extracting the start date here? If yes, can we just check the date type like this?
```suggestion
            if block.get('date_type') == 'course-start-date':
                course_date_list[0]['date'] = block_date
```

Reply:
We are actually getting nearest dates from today date. One past nearest date from today and one future nearest date from today. We are not exactly extracting the start date for first index of dates list. It could be course start date but will not every time.